bakery/checkers/checkers.go

```diff
@@ -4,192 +4,187 @@ package checkers
-	// are no other checkers. This also prevents someone
-	// from inadvertently overriding the error condition.
-	if cond == CondError {
-		return errBadCaveat
-	}
-	checked := false
+// Info returns information on all the registered checkers, sorted by namespace
+// and then name.
+func (c *Checker) Info() []CheckerInfo {
+	checkers := make([]CheckerInfo, 0, len(c.checkers))
 	for _, c := range c.checkers {
-		checkerCond := c.Condition()
-		if checkerCond != "" && checkerCond != cond {
-			continue
-		}
-		if err := c.Check(cond, arg); err != nil {
-			if checkerCond == "" && errgo.Cause(err) == ErrCaveatNotRecognized {
-				continue
-			}
-			return errgo.Mask(err, errgo.Any)
-		}
-		checked = true
+		checkers = append(checkers, c)
 	}
-	if !checked {
-		return ErrCaveatNotRecognized
-	}
-	return nil
+	sort.Sort(checkerInfoByName(checkers))
+	return checkers
 }
 
-// Condition implements Checker.Condition.
-func (c *MultiChecker) Condition() string {
-	return ""
+// Namespace returns the namespace associated with the
+// checker. It implements bakery.FirstPartyCaveatChecker.Namespace.
+func (c *Checker) Namespace() *Namespace {
+	return c.namespace
 }
 
-// CheckFirstPartyCaveat implements bakery.FirstPartyChecker.CheckFirstPartyCaveat.
-func (c *MultiChecker) CheckFirstPartyCaveat(cav string) error {
+// CheckFirstPartyCaveat implements bakery.FirstPartyCaveatChecker
+// by checking the caveat against all registered caveats conditions.
+func (c *Checker) CheckFirstPartyCaveat(ctxt context.Context, cav string) error {
 	cond, arg, err := ParseCaveat(cav)
 	if err != nil {
 		// If we can't parse it, perhaps it's in some other format,
 		// return a not-recognised error.
 		return errgo.WithCausef(err, ErrCaveatNotRecognized, "cannot parse caveat %q", cav)
 	}
-	if err := c.Check(cond, arg); err != nil {
+	cf, ok := c.checkers[cond]
+	if !ok {
+		return errgo.NoteMask(ErrCaveatNotRecognized, fmt.Sprintf("caveat %q not satisfied", cav), errgo.Any)
+	}
+	if err := cf.Check(ctxt, cond, arg); err != nil {
 		return errgo.NoteMask(err, fmt.Sprintf("caveat %q not satisfied", cav), errgo.Any)
 	}
 	return nil
 }
 
-// TODO add multiChecker.CheckThirdPartyCaveat ?
-// i.e. make this stuff reusable for 3rd party caveats too.
-
-func firstParty(cond, arg string) Caveat {
-	return Caveat{
-		Condition: cond + " " + arg,
-	}
-}
-
-// CheckerFunc implements Checker for a function.
-type CheckerFunc struct {
-	// Condition_ holds the condition that the checker
-	// implements.
-	Condition_ string
-
-	// Check_ holds the function to call to make the check.
-	Check_ func(cond, arg string) error
-}
-
-// Condition implements Checker.Condition.
-func (f CheckerFunc) Condition() string {
-	return f.Condition_
-}
+var errBadCaveat = errgo.New("bad caveat")
 
-// Check implements Checker.Check
-func (f CheckerFunc) Check(cond, arg string) error {
-	return f.Check_(cond, arg)
+func checkError(ctxt context.Context, _, arg string) error {
```

Comment:
Yeah, I can see your reasons. I think (if it's OK) that I'll leave code movement alone for the moment - I think it might be worth moving the Checker code into its own file and having checkers.go for checker implementations only, but I've got a PR after this one that gives me conflicts on every change, so I'd prefer to wait for that if that's OK.